compiler/rustc_mir/src/dataflow/impls/mod.rs

```diff
@@ -715,3 +669,82 @@ fn switch_on_enum_discriminant(
+
+    let (enum_place, enum_def) = match enum_ {
+        Some(x) => x,
+        None => return,
+    };
+
+    let enum_mpi = match move_data.rev_lookup.find(enum_place.as_ref()) {
+        LookupResult::Exact(mpi) => mpi,
+        LookupResult::Parent(_) => return,
+    };
+
+    let enum_path = &move_data.move_paths[enum_mpi];
+    let mut discriminants = enum_def.discriminants(tcx);
+
+    // `MovePathIndex`s for those variants with their own outgoing edge. These
+    // get marked as uninitialized along the "otherwise" edge.
+    let mut variant_mpis_with_edge = vec![];
+
+    edge_effects.apply(|trans, edge| {
+        if let Some(value) = edge.value {
+            // MIR building adds discriminants to the `values` array in the same order as they
+            // are yielded by `AdtDef::discriminants`. We rely on this to match each
+            // discriminant in `values` to its corresponding variant in linear time.
+            let (active_variant, _) = discriminants
+                .find(|&(_, discr)| discr.val == value)
+                .expect("Order of `AdtDef::discriminants` differed from `SwitchInt::values`");
+
+            for (variant_mpi, variant_path) in enum_path.children(&move_data.move_paths) {
+                // Because of the way we build the `MoveData` tree, each child should have exactly one more
+                // projection than `enum_place`. This additional projection must be a downcast since the
+                // base is an enum.
+                let (downcast, base_proj) = variant_path.place.projection.split_last().unwrap();
+                assert_eq!(enum_place.projection.len(), base_proj.len());
+
+                let variant_idx = match *downcast {
+                    mir::ProjectionElem::Downcast(_, idx) => idx,
+                    _ => unreachable!(),
+                };
+
+                if variant_idx == active_variant {
+                    // If this is the move path for the variant that is active along this edge
+                    // of the `SwitchInt` terminator, remember that it is not active as part of
+                    // the "otherwise" edge.
+                    variant_mpis_with_edge.push(variant_mpi);
+                } else {
+                    // Otherwise, it is the move path for an *inactive* variant. Mark it and
+                    // all of its descendants as uninitialzied.
+                    drop_flag_effects::on_all_children_bits(
+                        tcx,
+                        body,
+                        move_data,
+                        variant_mpi,
+                        |mpi| on_uninitialized_variant(trans, mpi),
+                    );
+                }
+            }
+        } else {
+            // We're on the otherwise branch. All variants that we visited above are known to be
+            // uninitialized.
+            for &variant_mpi in &variant_mpis_with_edge {
```

Comment:
Is there a guarantee that the otherwise branch is always invoked *last* during the iteration performed by `edge_effects.apply`?

Because, as best as I can tell, it seems like you are relying here on such a guarantee (since you are accumulating all the values in `variant_mpis_with_edge` during the callback invocations on the non-otherwise edges.

But skimming over the doc comments added in PR #77242, I do not see any such guarantee in the documentation for `apply_switch_int_edge_effects`.

(And overall, I think the logic here would be easier to understand if the block structure in the code reflected that ordering constraint, even if it requires changes to the dataflow API. I don't know the best way to express it; perhaps by two distinct callbacks instead of just one, where the first callback is invoked for all the `edge.value.is_some()` cases here, and the second callback for the otherwise case?)